### PR TITLE
update the docs for layout guarantees of option-like enums

### PR DESCRIPTION
**Pull request by RalfJung**
This updates our docs to match the FCP in https://github.com/rust-lang/rust/pull/130628. The UCG reference has been a valuable resource for figuring out what is and is not guaranteed about layout, so it seems worth keeping it up-to-date.

(Medium-term, I expect this to be superseded by @chorman0773's new layout chapter for the reference, but we're not there yet.)

### CODE REVIEW

**Comment by RalfJung** [comment]
@rust-lang/opsem could someone take a look at this so we can get it merged? :)

File: reference/src/layout/enums.md
Comment (saethlin) [inline reference/src/layout/enums.md:306]: Restricting the definition of an option-like enum based on the type of the single field is quite strange to me. Because I'd think that excludes `Option<T>` from the definition, because `T` could be `usize` or any type without guaranteed discriminant elision.

Comment (RalfJung) [reply reference/src/layout/enums.md:306]: @saethlin does this answer your question? I don't think I quite understood the question, TBH.

Comment (saethlin) [reply reference/src/layout/enums.md:306]: Whoops, I didn't notice you'd responded.

The text here is just confusing to me, because what I'm imagining is a reader saying "Is this an option-like enum":
```rust
enum Thing<T> {
    Yes(T),
    No,
}
```
Because by the text above, `Thing<T>` is not "an option-like enum" because its single field does not guarantee discriminant elision. This would be though:
```rust
enum Thing {
    Yes(NonZeroUsize),
    No,
}
```

...but *directly below this* the reference already says that `Option<T>` is an example of an option-like enum.

:shrug: 

I think you're making an assumption that the reader will interpret words a bit differently when a generic parameter is included. If other can confirm that they have read this and don't find the wording confusing I'm happy to be overruled.

Comment (chorman0773) [reply reference/src/layout/enums.md:306]: For this reason, in my type-layout docs PR, I chose to separate the idea of "Is an enum for which discriminant elision may apply" (in my text, called a *discriminant elision elible*)  and "A field of this type is subject to discriminant elision" (called *elision candidate type*). When the two are combined is when the result is guaranteed.
I also note explicitly that the determiniation is post-mono. and explicitly call out both `Option<T>` and `Result<T,E>`.

Comment (RalfJung) [reply reference/src/layout/enums.md:306]: I am implicitly assuming that we are talking about a fully monomorphic type here. It doesn't make sense to ask about the layout of a generic type. I can make that more explicit (though this will come at the cost of readability). I don't see how splitting this concept into two separate concepts is helpful.

**Comment by saethlin** [comment]
The definition at the top is at least something to point to. I'm still not enthused about the wording, but I think it's better to merge this and wait for someone to actually be confused than continue to haggle over it.

**Review by saethlin — APPROVED** [review]

**Comment by saethlin** [comment]
I don't know which of the merge buttons you want to use here, so I'll let you merge this :)

**Comment by RalfJung** [comment]
I see only one merge button?^^ :)

**Comment by ia0** [comment]
I guess the question was between merge, rebase, and squash (all sharing the same button using a drop down for selection). But given the git history and how this PR was merged, the answer was indeed merge 😅 This can be enforced with a branch rule if needed, such that only one option is available in the drop down, making such question automatically answered in the future.

**Pull request merged**
